nomad/encrypter.go

```diff
@@ -434,35 +435,42 @@ func (krr *KeyringReplicator) stop() {
 const keyringReplicationRate = 10
 
 func (krr *KeyringReplicator) run(ctx context.Context) {
-	limiter := rate.NewLimiter(keyringReplicationRate, keyringReplicationRate)
 	krr.logger.Debug("starting encryption key replication")
 	defer krr.logger.Debug("exiting key replication")
 
-	retryErrTimer, stop := helper.NewSafeTimer(time.Second * 1)
-	defer stop()
-
-START:
 	store := krr.srv.fsm.State()
+	abandonCh := store.AbandonCh()
+	minIndex := uint64(0)
 
 	for {
 		select {
 		case <-krr.srv.shutdownCtx.Done():
 			return
 		case <-ctx.Done():
 			return
+		case <-abandonCh:
+			// reset the blocking query to make sure we get all the keys from a
+			// restored snapshot
+			store = krr.srv.fsm.State()
+			abandonCh = store.AbandonCh()
+			minIndex = uint64(0)
 		default:
-			// Rate limit how often we attempt replication
-			err := limiter.Wait(ctx)
+			// ensure that we abandon the blocking query periodically to give
+			// other contexts a chance to fire
+			queryCtx, queryCancel := context.WithTimeout(ctx, time.Second)
+			defer queryCancel()
+
+			var rawIter any
+			var err error
+			rawIter, minIndex, err = store.BlockingQuery(getRootKeyMetas, minIndex, queryCtx)
```

Comment:
Hum..just thought of something, but I'm not sure if this how the blocking query works. 

Instead of a timeout could we instead check `abandonCh` after the query returns and loop back if the query returned because the snapshot was abandoned? So something like:

```go
for {
  select {
    // ...
    default: 
      // ...
      rawIter, minIndex, err = store.BlockingQuery(getRootKeyMetas, minIndex, ctx)
      select {
        case <-abandonCh:
          store = krr.srv.fsm.State()
          // ...
          continue 
        default:
      }
      // ...
}
```

Reply:
Oh too bad. I thought the problem was that we were getting out of the blocking query too early 😅

Reply:
I think there's either some very wrong with my implementation of the blocking query or there's a bug in `BlockingQuery`, because that's generating _many many_ queries rather than waiting on the wait channel as I'd expect. That in turn seems to be destableizing leadership in the unit tests.

In the interest of shipping this critical fix sooner rather than later, I've backed out the blocking query in lieu of a rate-limited one, and that seems to have stabilized the tests. Lemme do a little bit of cleanup of the resulting code and I'll flag this for review one last time.

Reply:
What's really weird about this is that the timeout we implemented makes this very similar to the `blockingRPC` function, which sets a context that the `ws.WatchCtx` uses.